Fastsync English spec for review

josef-widder opened the pull request.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+# Fastsync
+
+<!--
```
Maybe we should remove these comments from the template.

josef-widder replied:
I removed them in a previous revision

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+
+In practice it is sufficient to catch-up only close to the current
+height: The Tendermint consensus reactor implements a similar
+functionality and can synchronize a node that is approximately 100
```
I don't remember seeing code that does this in consensus reactor. @ebuchman @melekes Can someone confirm this? Ideally we should also add a reference to the corresponding code here.

melekes replied:
Consensus reactor can catch up, but it's very slow comparing to fast sync & state sync.

josef-widder replied:
Ethan changed it to 10 already.

andrey-kuprianov commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+
+#### **[TMBC-CorrFull]**: 
+Every correct full node locally stores a prefix of the
+current list of headers from [**[TMBC-SEQ]**](TMBC-SEQ-link).
```
```suggestion
current list of headers from [**[TMBC-SEQ]**][TMBC-SEQ-link].
```
the link is broken

ebuchman commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+is greater than the time it takes to add a block to the blockchain
+(LTIME in [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]), the
+protocol may never terminate and thus violate [FS-VC-TERM].  This
+scenario is even possible if a correct peer is always in *peerIDs*,
```
Slash isn't it possible with just one correct peer if we manage to always eventually request pairs of sequential blocks from that correct peer?

milosevic replied:
I guess your comment applies to the case above: we could ask always consecutive blocks from a peer, but it would lead to waste of resources as you would need for every block to download two blocks (in case requests are made to different peers). More efficient solution would be asking block and commit from a single peer, but that would require Tendermint API.

ebuchman commented:
Other high level thoughts I'll try to summarize in advance of a discussion:

- Is there convention on the definition names over what gets upper cased and what doesnt? Eg. SEQ vs Auth ...
- Can we include some kind of legend to make the short forms more understandable, ie:
  - FM = Fault Model
  - A = Assumption
  - FS = Fast Sync
  - etc ...
- Without the context of the VDD document, or even with it, some of the titles/headings are a bit confusing. For instance the Sequential vs Distributed problem statement and other document organization. We can do a better job with a table of contents and description of how the doc is laid out and how to follow it.
- In our contract we distinguish between the "core" and "peer" state machines of a reactor. This is most notable in the consensus reactor, but we should make some effort to distinguish them here as well. For the fast sync, it's practically all "peer" state machine, though the verifycommit/execute logic would be the "core" state machine.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+
+#### **[TMBC-VALIDATOR-Set]**:
+
+A *validator set* is a set of validator pairs. For a validator set
```
The problem was, that in the existing documentation, "validator" refers to both a data structure and a full node. I added the definition. Validator pair refers to the data structure (Address + voting power) and validator node is the full node.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+
+*Fastsync* gets as input a block of height *h* and the corresponding
+application state *s* that corresponds to the block and state of that
+height of the blockchain [**[TMBC-SEQ]**][TMBC-SEQ-link], and produces
```
I guess it is not. I removed the reference.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+as output (i) a list *L* of blocks starting at height *h* to some height
+*terminationHeight*, and (ii) the application state when applying the
+transaction of the list *L* to *s*. Fastsync has to satisfy the following
+properties [FS-Seq-?]:
```
I wanted to use '?' as wildcard but I guess it is not clear. Any idea how to write that?

konnov replied:
Once we have traceability identifiers around, patterns like `?` will be unparseable by the tool.

josef-widder replied:
Zarko suggested to remove that tag

ebuchman replied:
Ah, I'm used to * as a wildcard, but in any case think it can just be removed here

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+#### **[FS-Seq-Term]**:
+Let *bh* be the height of the blockchain at the time *Fastsync* starts.
+When *Fastsync* terminates, it outputs a list of all blocks from
+height *h* to some height *terminationHeight >= bh - 1*,
```
Yes. Good point. I added a clarification right there.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+Let *bh* be the height of the blockchain at the time *Fastsync* starts.
+When *Fastsync* terminates, it outputs a list of all blocks from
+height *h* to some height *terminationHeight >= bh - 1*,
+[**[TMBC-SEQ]**][TMBC-SEQ-link].
```
I agree. It is not so clear (and not consistently throughout the document). I removed the references where they were unclear.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+> While in principle the above condition can be part of a sufficient
+> condition to solve [FS-Seq-Live] and
+> [FS-Seq-Term] and [FS-Seq-Term-SYNC], we will see below that the
+> current implementation of Fastsync (V2) requires the following (much
```
As far as I understand the problem (or a similar one) also is present at the other implementations. We should discuss what precisely we can put here.

milosevic replied:
It is also true for the past versions.

josef-widder commented on docs/spec/fastsync/fastsync.md:
````diff
@@ -0,0 +1,1101 @@
+```
+- Comment
+    - *ht* is a height
+	- peers can provide the status without being called
````
I agree. I aimed at a more uniform terminology with e.g. the light client specs, and at some point I realized I needed that particular addition. The nice thing about functions is that I can describe expected pre- and postconditions very nicely. Let's discuss how to express what we need.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+- Expected precondition
+	- application state is the one of the blockchain at height
+      *height - 1*
+	- **[FS-V2-Verif]** for any two blocks *a* and *b* from
```
No, I want this particular condition to be labeled for later reference.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+    - *peerHeights(addr) = ht*
+	- *TargetHeight* is updated
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs* (that is,
```
It means the latter, that is, "if it returns an error, then addr will be removed from peerIDs"

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+- Expected postcondition
+    - if function `Execute` has been executed without error or was not executed:
+        - *receivedBlocks(b.Height) = addr*
+	    - *blockstore(b.Height) = b*
```
Yes, this just describes bookkeeping that is in addition to what is done within Execute.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+- *startHeight = startBlock.Height*
+
+#### Additional Variables
+- *trustedBlockstore*: stores for each height greater than or equal to
```
Do you mean to rename it?

josef-widder commented on docs/spec/fastsync/fastsync.md:
````diff
@@ -0,0 +1,1101 @@
+
+
+```go
+func SequentialVerify {
````
We can discuss in what detail we want to describe possible fixes.

milosevic replied:
In my view, it is sufficient to describe the fix at the high level, as done above, so we don't need to enter the details of the pseudocode.

josef-widder commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+	- [NewFS-VAR-TRUST-INV]
+- Expected postcondition
+    - [NewFS-VAR-TRUST-INV]
+	- there is no block *bnew* with *bnew.Height = height + 1* in
```
This says that you don't stop as long as there is another block in blockstore, or in other words, you iterate to the maximal possible height at that moment.

josef-widder commented:
> * Is there convention on the definition names over what gets upper cased and what doesnt? Eg. SEQ vs Auth ...

No. We will fix a convention soon. For now, I changed all to upper case.

> * Can we include some kind of legend to make the short forms more understandable, ie:
>   
>   * FM = Fault Model
>   * A = Assumption
>   * FS = Fast Sync
>   * etc ...

I added the most frequently used ones

> * Without the context of the VDD document, or even with it, some of the titles/headings are a bit confusing. For instance the Sequential vs Distributed problem statement and other document organization. We can do a better job with a table of contents and description of how the doc is laid out and how to follow it.

I added added a bullet list at the top of the document with pointers to the sections. Please, let me know what you think!

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+
+
+#### **[TMBC-AUTH-BYZ]**:
+The authenticated Byzantine model assumes that no node (faulty or
```
```suggestion
We assume the authenticated Byzantine model in which no node (faulty or
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+#### **[TMBC-AUTH-BYZ]**:
+The authenticated Byzantine model assumes that no node (faulty or
+correct) may break digital signatures, but otherwise, no additional
+assumption is made about the internal behavior of faulty full
```
Full node is not defined at this point in time. Maybe it's sufficient to say here just faulty nodes.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+<!-- cannot forge messages [TMBC-Sign-NoForge]. -->
+
+
+> We observed that in the existing documentation the term
```
What documentation? Tendermint documentation?

josef-widder replied:
yes.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+*validator pair* is a pair *(address, voting_power)*, where 
+  - *address* is the address (public key) of a full node, 
+  - *voting_power* is an integer (representing the full node's
+  voting power in a certain consensus instance).
```
```suggestion
  voting power in a given consensus instance).
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+  voting power in a certain consensus instance).
+  
+> In the Golang implementation the data type for *validator
+> pair* is called `Validator`
```
```suggestion
> pair* is called `Validator`.
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+
+#### **[TMBC-TIME-PARAMS]**:
+A blockchain has the following configuration parameters:
+ - *unbondingPeriod*: a time duration.
```
It seems that we don't use unbonding period term.

josef-widder replied:
I think we should still use it as people know the term?

milosevic replied:
Ok, fair enough.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+
+
+
+## Informal Problem statement
```
```suggestion
## Fastsync Informal Problem statement
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+A full node has as input a block of the blockchain at height *h* and
+the corresponding application state (or the prefix of the current
+blockchain until height *h*). It has access to a set *peerIDs* of full
+nodes called *peers* that it knows of.  The full node uses the peers
```
Full node is never defined if I am not wrong.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+as output (i) a list *L* of blocks starting at height *h* to some height
+*terminationHeight*, and (ii) the application state when applying the
+transactions of the list *L* to *s*. Fastsync has to satisfy the following
+properties [FS-SEQ-?]:
```
```suggestion
properties:
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+- FS: Fastsync
+- LIVE: liveness
+- INV: invariant
+- TERM: termination
```
What is difference between liveness and termination? I would expect that it is the same and that one abbreviation is sufficient. On the other side we don't have abbreviation for safety properties.

josef-widder replied:
As you noted below, TERM was mostly addressing safety properties upon termination. I will remove TERM here, as we have changed the tags

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+Let *bh* be the height of the blockchain at the time *Fastsync*
+starts. By assumption we have *bh >= h*.
+When *Fastsync* terminates, it outputs a list of all blocks from
+height *h* to some height *terminationHeight >= bh - 1*.
```
I don't get this property. Not sure why you need to use both bh and h as starting height.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+> blockchain while Fastsync is running. It links the target height to the
+> initial state. If Fastsync has to catch-up many blocks, it would be
+> better to link the target height to a time close to the
+> termination. This is capture by the following specification:
```
```suggestion
> termination. This is captured by the following specification:
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+*Fastsync* eventually terminates.
+
+ 
+#### **[FS-SEQ-TERM]**:
```
This is safety property, so the tag could probably be FS-SAFE-START.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+> better to link the target height to a time close to the
+> termination. This is capture by the following specification:
+
+#### **[FS-SEQ-TERM-SYNC]**:
```
FS-SAFE-SYNC?

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+the blockchain at height *terminationHeight*.
+
+> As the block of height *h + 1* is needed to verify the block of
+> height *h* we highlight th following clarification of the
```
```suggestion
> height *h* we highlight the following clarification of the
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+#### **[FS-SEQ-HEIGHT]**: 
+The returned value *terminationHeight* is the height of the block with the largest
+height that could be verified. In order to do so, *Fastsync* needs the
+Commit of the block at height  *terminationHeight + 1* in the blockchain.
```
```suggestion the block at height  *terminationHeight + 1* in the blockchain.
```

milosevic replied:
```suggestion
the block at height  *terminationHeight + 1* in the blockchain.
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+
+#### **[FS-A-PEER-IDS]**:
+*FS* has access to a set *peerIDs* of IDs (public keys) of peers (full
+     nodes). During the execution of *Fastsync*, another protocol (outside
```
```suggestion
     . During the execution of *Fastsync*, another protocol (outside
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+We consider a node *FS* that performs *Fastsync*.
+
+#### **[FS-A-PEER-IDS]**:
+*FS* has access to a set *peerIDs* of IDs (public keys) of peers (full
```
```suggestion
*FS* has access to a set *peerIDs* of IDs (public keys) of peers
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+according to [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link].
+
+#### **[FS-A-VAL]**:
+The system satisfies [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link] and [**[TMBC-FM-2THIRDS]**][TMBC-FM-2THIRDS-link]. Thus, there is a
```
This is not really clear in my view. We have a node and peers, and I guess that with this property we want to make a link to validators and Tendermint blockchain. I am not sure what is the easiest way to introduce this without going into details of explaining validators and difference between validators and full nodes. From the fast sync view point the whole blockchain business can be seen as application specific stuff, and it is only relevant as part of block validation in the proposed solution.

josef-widder replied:
If we want to say what blocks to accept and execute, we have to state that there is a reference block chain that has these blocks. It is not clear how to specify reasonable invariants without having that reference. It is the validity property of Fastsync.   
Later, in our suggestions to use the lightclient-like verification (at the end of the document) we need this as it is based on the tendermint failure model.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+
+#### Remote Functions
+
+The Tendermint Full Node exposes the following functions over
```
We haven't defined what is a Tendermint full node. Can we just say that peers expose the following functions over RPC?

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+> stronger) requirement
+
+#### **[FS-ALL-CORR-PEER]**:
+At all times, the set *peerIDs* contains only correct full nodes.
```
We need to check if some properties of v2 hold in the crash tolerant model.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+
+
+#### **[FS-VC-STATE-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
```
Isn't this always the case? Difference between success and failure in my view is only with respect to the quality of the result, i.e, about number of synced blocks.

josef-widder replied:
I wanted to recall *terminationheight* here, but we can do without, I removed it.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+height *terminationHeight*.
+
+#### **[FS-VC-BLOCKS-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
```
The same comment as above.

josef-widder replied:
removed it

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+Under [FS-ALL-CORR-PEER], let *maxh* be the maximum 
+height of a correct peer [**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link]
+in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates
+successfully, it is at some height *terminationHeight >= maxh - 1*.
```
I might be wrong but I haven't seen definition of termination height. Is that the height of the last verified and executed block?

josef-widder replied:
you are wrong ;-) we have the following:
#### **[FS-SEQ-HEIGHT]**: 
The returned value *terminationHeight* is the height of the block with the largest
height that could be verified. In order to do so, *Fastsync* needs the
block at height  *terminationHeight + 1* of the blockchain.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates
+successfully, it is at some height *terminationHeight >= maxh - 1*.
+
+> The above property is independent of how many blocks are added to the
```
It is also independent of addition of new peers. As peers can be removed by external component (not just added), we probably need additional assumption that a correct peer is not removed from the peerIds by the external component.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1164 @@
+that if *term* is the time *Fastsync* terminates and
+*maxh* be the maximum height of a correct peer
+[**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link] in *peerIDs* at the time
+*term - TD*, then if *FastSync* terminates successfully, it is at
```
Sequential property (FS-SEQ-TERM-SYNC) refers to a logical diff (parameter D that is expressed as lag in number of heights), while here we introduce something similar but not the same with time parameter TD. Can we somehow align this better or it is fine as sequential spec is defined without reference to system parameters.

josef-widder replied:
I agree the issue here is not explicitly discussed. The problem is that in the case the peers are themselves synchronized to the blockchain (they are at the current height), then *TD* together with LTIME in [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]  can be used to bound $D$. If the peers themselves lag behind and catch-up during the *TD* time, we cannon bound anything. Perhaps it is enough to add an explanation at this point to make it clear. What do you think?

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+---->
+
+#### **[FS-VC-ALL-CORR-TERM]**:
+Under [FS-ALL-CORR-PEER], *Fastsync* eventually terminates successfully.
```
Where do we define what we mean by termination with failure?

josef-widder replied:
In the following. Is this OK?

#### **[FS-V2-TIMEOUT]**:

*Fastsync V2* starts a timeout whenever a block is
executed (that is, when the height is incremented). If the timeout expires
before the next block is executed, *Fastsync* terminates.
We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
with failure, otherwise it terminates successfully.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+- *blockstore*: stores for each height greater than
+    *startBlock.Height*, the block of that height. initially nil for
+    all heights
+- *peerTimeStamp*: stores for each peer the last time a block was received
```
There is also pendingTime variable in the code where time of the block request is stored. It is used as part of tryPrune function but only for the peer that is responsible for current height.

josef-widder replied:
I added at this point

- *pendingTime*: stores for each peer the last time a block was requested

and later, I updated:

#### **[FS-V2-PEER-REMOVE]**:
Periodically, *peerTimeStamp* and *peerRate* and *pendingTime* are
analyzed. 
If a peer *p*
has not provided a block recently (check of *peerTimeStamp[p]*) or it
has not provided sufficiently many data (check of *peerRate[p]*), then
*p* is removed from *peerIDs*. In addition, *pendingTime* is used to
estimate whether the peer that is responsible for current height is
still connected.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+---->
+
+The protocol is described in terms of functions that are triggered by
+(external) events. The implementation uses a scheduler and a
```
I am not sure that this sentence is correct. Maybe we just need to explain that some messages are sent periodically (statusRequest) and some actions that lead to sending messages are also triggered periodically. We just need to explain basically execution model.

josef-widder replied:
At some point we said we want to keep this underspecified on purpose, now it seems that this and the other comments target at specifying V2 more precisely. We should fix what level of abstraction we are aiming for.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+  one from a peer. It does so by calling `Block(n,h)` remotely on one
+  peer *n* for a missing height *h*.
+  
+> We have left the strategy how peers are selected unspecified, and
```
Maybe we should explain the strategy used as it is part of the solution. It basically selects a peer with the minimum number of pending requests that can serve given height h (peerHeight >= h).

josef-widder replied:
I added the clarification for V2 to the comment.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+  
+#### **[FS-V2-TIMEOUT]**:
+
+*Fastsync V2* starts a timeout whenever a block is
```
To be precise, implementation is slightly different. Node remember the last point in time when a block has been executed. Then there is an action that is triggered periodically and if no progress has been made for at least syncTimeout, i.e., (now - lastUpdate > syncTimeout) then fastSync terminates.

josef-widder replied:
We can describe it more precisely, but I would consider this an implementation detail of how it is checked whether a timeout occurred.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+*Fastsync V2* starts a timeout whenever a block is
+executed (that is, when the height is incremented). If the timeout expires
+before the next block is executed, *Fastsync* terminates.
+We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
```
Not sure about this. Note that you can terminate in this case with a peerset not empty, and I am not sure if it is still considered success. My understanding was that success is in case we sync up to maxHeight and failure is when we terminate because of this timeout.

josef-widder replied:
I thought that in the synchronous system your scenario is not possible. But I guess I didn't make explicit that the timeout must be sufficiently large.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+    - if precondition is violated: *addr* not in *peerIDs* (that is,
+      *addr* is removed from *peerIDs*)
+- Timeout condition
+    - if `OnStatusResponse(addr, ht)` was not invoked within *2 Delta* after
```
Note that in V2 this is not the case, i.e., it is not tracked when statusResponse arrived. Therefore, peers are not removed upon late statusReply.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+- Expected postcondition
+    - Function `Block` is called remotely at a peer *addr* in peerIDs 
+	  for a missing height *h*  
+	  *Remark:* different implementations may have different
```
As this spec is about v2, we should probably details strategy implemented in v2.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+	  for a missing height *h*  
+	  *Remark:* different implementations may have different
+      strategies to balance the load over the peers
+    - *pendingblocks(h) = addr*
```
It also remember time at which request is made, but don't know if we want to enter that level of details as you capture it at more high level above.

milosevic commented on docs/spec/fastsync/fastsync.md:
````diff
@@ -0,0 +1,1163 @@
+```
+- Comment
+    - if after adding block *b*, blocks of heights *height* and
+      *height + 1* are in *blockstore*, then `Execute` is called
````
Execute is in v2 triggered periodically, i.e., it is not triggered by block reception.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+      *height + 1* are in *blockstore*, then `Execute` is called
+- Expected precondition
+    - *pendingblocks(b.Height) = addr*
+	- *b* satisfies basic soundness  
```
blockstore(b.Height) = nil (empty)

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+      executed:
+        - *receivedBlocks(b.Height) = addr*
+        - *blockstore(b.Height) = b*
+        - *peerTimeStamp[addr]* is set to a time between invocation and
```
peerTimestamp[addr] is set to the current time in v2, but maybe you want to capture here more high level concept.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+        - *peerTimeStamp[addr]* is set to a time between invocation and
+          return of the function.
+        - *peerRate[addr]* is updated according to size of received
+          block
```
```suggestion
          block and time it has passed between current time and last block received from this peer (addr)
```

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+        - *peerRate[addr]* is updated according to size of received
+          block
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs*; reset
```
Removing a peer triggers cleaning data structure such that all pending/received but not processed requests/blocks are removed.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+    - if precondition is violated: *addr* not in *peerIDs*; reset
+	*pendingblocks(b.Height)* to nil;
+- Timeout condition
+    - if `OnBlockResponse(addr, b)` was not invoked within *2 Delta* after
```
In v2, there is no such check. Removal happens periodically but only for the current height this kind of check is implemented. For other peers, it looks at peer rate and difference between current time and last block reception.

milosevic commented on docs/spec/fastsync/fastsync.md:
````diff
@@ -0,0 +1,1163 @@
+```
+- Comments
+    - none
+- Expected precondition
````
Precondition can be that blocks for height and height+1 are in the blockStore, and that VerifyCommit (a,b.Commit) = true. Postcondition is that block at height is executed, state updated and height incremented. There are two error cases: 1) blocks are not present, do nothing, or 2) blocks are present but VerifyCommit (a,b.Commit) = false. Then both blocks are removed and the corresponding peers are removed.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+*peerIDs*. If block *a* was correct and provided by a correct peer *p*,
+and block b was faulty and provided by a faulty peer, the protocol
+- removes the correct peer *p*, although it might be useful to
+  download blocks from it in the future
```
Maybe it's better to link it with failed invariant (we don't download all blocks from correct peers).

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1101 @@
+			exit;
+		}
+		if ValidCommit(trustedBlockstore[height - 1].NextValidators, b2.commit) {
+			// we trust b2
```
This follows from the definition above.

milosevic commented on docs/spec/fastsync/fastsync.md:
```diff
@@ -0,0 +1,1163 @@
+   assumes that an arbitrary number of blocks can be obtained and
+   checked within one round-trip 2 Delta; which either needs
+   conservative estimation of Delta, or a more refined analysis). Thus
+   we reach the *targetHeight* and terminate by time *end*.
```
I like this analysis. This was my intuition, but I wasn't able to formalise it. I am just wondering if faulty peers sending statusResponse for nonexistent blocks can prevent us from termination. I am wondering also if the fact that correct peers are not fully up to date with the blockchain matters. It seems that what this proposal ensures is that once we hit max height of correct peers, if faulty peers advertise higher height they will eventually be removed as *2 Delta < ETIME*.
